vm/src/stdlib/io.rs

```diff
@@ -2202,10 +2202,26 @@ mod _io {
             *data = None;
 
             let encoding = match args.encoding {
-                Some(enc) => enc,
+                Some(enc) => {
+                    if enc.as_str() == "locale" {
+                        vm.import("locale", None, 0)?
+                            .get_attr("getencoding", vm)?
+                            .call((), vm)?
+                            .try_into_value(vm)?
+                    } else {
+                        enc
+                    }
+                }
                 None => {
-                    // TODO: try os.device_encoding(fileno) and then locale.getpreferredencoding()
-                    PyStr::from(crate::codecs::DEFAULT_ENCODING).into_ref(&vm.ctx)
+                    if vm.state.settings.utf8_mode == 1 {
+                        // Used to be crate::codecs::DEFAULT_ENCODING, which was "utf-8"
+                        PyStr::from("utf-8").into_ref(&vm.ctx)
+                    } else {
+                        vm.import("locale", None, 0)?
+                            .get_attr("getencoding", vm)?
+                            .call((), vm)?
+                            .try_into_value(vm)?
+                    }
```

Comment:
conditional match arms will be useful.
Also, we can directly use `crate::codecs::DEFAULT_ENCODING` rather than copying the value.
```suggestion
                None if vm.state.settings.utf8_mode == 1 => {
                    PyStr::from(crate::codecs::DEFAULT_ENCODING).into_ref(&vm.ctx)
                }
                Some(enc) if enc.as_str() != "locale" => enc,
                _ => {
                    // None without utf8_mode or "locale" encoding
                    vm.import("locale", None, 0)?
                        .get_attr("getencoding", vm)?
                        .call((), vm)?
                        .try_into_value(vm)?
                }
```

Reply:
Yeah, I decided to put "utf-8" instead of default encoding because the corresponding cpython code explicitly called for utf-8, not default encoding. Those two are same atm for rustpython, but I figured that kind of implicit assumption are better removed.

Ah, I forgot that I could use match guards to further refine match cases. 

I think that is a very nice way to put it. At first I did not get why you used a default case rather than a case without a match guard, but then I realized that that would again lead to code duplication, which was the very issue that you tried to solve.

Pushed!